torch/_functorch/_activation_checkpointing/ac_reorder_inference_pass.py

```diff
@@ -0,0 +1,143 @@
+"""
+AC reordering pass: Duplicates checkpointed nodes for backward, then DCE removes unused forward versions.
+"""
+
+import torch
+import torch.fx as fx
+from torch._functorch import config
+from torch._functorch.compile_utils import raise_getitems
+from torch._functorch.partitioners import (
+    cleanup_recompute_tags,
+    force_save_bw_mutation_src,
+    force_save_collectives,
+    has_recomputable_ops,
+    has_recomputable_rng_ops,
+    is_not_collective,
+    must_recompute,
+)
+
+
+def is_impure_node_for_dce(node):
+    # Check for special collectives that should be treated as pure
+    if not is_not_collective(node):
+        # It's a collective (wait_tensor, all_gather_into_tensor, etc.)
+        # Treat as pure - can be eliminated if unused
+        return False
+
+    # For everything else, fall back to the DEFAULT logic
+    # This is what eliminate_dead_code() calls when is_impure_node=None
+    impure_random = True
+    if torch._guards.TracingContext.try_get():
+        impure_random = torch._inductor.config.fallback_random
+    return node.is_impure(impure_random)
+
+
+def _is_backward_node(node: fx.Node) -> bool:
+    """Check if node is in backward region via annotation"""
+    return node.meta.get("custom", {}).get("backward") is not None
+
+
+def rematerialize_nodes_with_ac_annotations(gm: fx.GraphModule) -> fx.GraphModule:
+    """
+    Duplicate checkpointed nodes for backward use. DCE removes unused forward versions.
+    """
+    # Early exit if no checkpointed ops
+    if not has_recomputable_ops(gm):
+        return gm
+
+    # Ban RNG ops in checkpointed regions (like partitioner does for inference mode)
+    # Note: RNG functionalization is not currently supported for inference mode graphs
+    if has_recomputable_rng_ops(gm):
+        raise RuntimeError(
+            "Activation checkpoint reordering in fullgraph does not support RNG ops in checkpointed regions. "
+            "RNG state cannot be properly synchronized between "
+            "forward and recomputed backward operations. Please move RNG operations outside "
+            "of checkpoint regions, or use joint graph mode (where partitioner handles RNG)."
```

Comment:
Some nits on this message:

I think we are generally referring to the inference graph / fullgraph as "forward-loss-backward graph" . Could we change this to something like "Activation checkpoint for forward-loss-backward graph does not currently support RNG ops"
The user probably doesn't need to know about the details like  "RNG state cannot be properly synchronized", an the workarounds are kind of obvious, we can probably leave them out lol